### ISSUE TITLE
Ability to modify transcoder arguments to utilize node capabilities

### ISSUE THREAD

**Issue by Camofelix** [pabloromeo/clusterplex#34]
TLDR; method and/or advice on how to implement argument substitution to take advantage of HW accels on worker node platforms. 

Hey Guys, gals, non-binary pals!

I'm working on a low power embedded cluster [(repository describing project here)](https://github.com/Camofelix/Jetson_ffmpeg_trancode_cluster)

I've forked the UnicornTranscoder project already, and found a way to insert/substitute custom arguments into the FFmpeg arguments to take advantage of certain hardware acceleration functions based on the available devices. unfortunately such a solution is rather inelegant from a management and orchestration standpoint. 

to mitigate this I was initially looking at the Kube_plex project, but this looks much more fit for purpose.

For now I've forked the repo and begun to look into how it deals with passing arguments, and if something simplistic like a case:switch statement could achieve what I'm looking for.

The first and easiest of changes would be to "always" insert -hwaccel auto to allow FFmpeg and it's build in platform parcer to accelerate a portion of the transcoding. secondary to that would be dealing with things such as vaapi (for Intel and AMD/ATI) and vdpau/CUDA for Nvidia discrete cards

Ideally this would be on the worker side and eventually have a configuration script select the relevant hwaccels available on platform such as OpenCL, AMF, VDPAU etc. 

Would love your thoughts! 

-FelixCLC

**Comment by Camofelix** [comment]
Started looking into how the project is architected. I *think*  that the easiest place to test this is by adding options via the "newargs " string at ~line 25-54 of /root_directory/pms/transcoder.js

**Comment by pabloromeo** [comment]
Hey!!
How's it going? Interesting use of Jetson!
So, if I understand correctly your plan would be to somehow override Plex's own version of ffmpeg (which in their installs is an executable called "Plex Transcoder" in /usr/lib/plexmediaserver/) with your own, right?
Now, beware that the version distributed with Plex is their own fork, with extra code. Not the standard ffmpeg. Not sure if there is any incompatibility in the arguments used (for example, the -progress-url argument added by Plex, to name one).

Clusterplex doesn't have its own custom transcoder, it's ultimately just calling Plex's, with the arguments forwarded from PMS to the remote Worker.

But now, there might actually be a way for you to customize arguments as you see fit, even without changing the existing code.
Meaning, if you look at the worker.js code, you'll see these two lines near the top:

```
const TRANSCODER_PATH = process.env.TRANSCODER_PATH || '/usr/lib/plexmediaserver/'
const TRANSCODER_NAME = process.env.TRANSCODER_NAME || 'Plex Transcoder'

```

So technically you can provide your own path and executable name for an ffmpeg transcoder, and the worker will call it, with the original arguments, environment variables, and current working directory value.

That means you can try one of three things:
1. Handle arguments as you see fit within your custom ffmpeg version and have those two environment variables point at your version.
2. If you want to use Plex's transcoder but want to manually alter arguments, you can have those two env variables point at an executable of your own making (like a bash script, for example) that alters arguments and then makes the exec call to `/usr/lib/plexmediaserver/Plex Transcoder`
3. A mix of both: Configure these two environment variables to point to a custom executable script that alters arguments, and then forwards the call to your own ffmpeg.

Disclaimer: I can't really test these alternatives at the moment, since I run this on a RPI4 cluster within docker swarm without any HW transcoding.

But forwarding the call to an intermediate script should work fine, as it is indeed the same approach used here with the `transcoder-shim.sh` that replaces Plex's own transcoder in PMS for Clusterplex's one that forwards the call to the Orchestrator.

Do you think that might be useful? At least for an initial test?

**Comment by Camofelix** [comment]
> So, if I understand correctly your plan would be to somehow override Plex's own version of ffmpeg (which in their installs is an executable called "Plex Transcoder" in /usr/lib/plexmediaserver/) with your own, right?

That's exactly correct! I'll be using the exact same "shim" method that most distributed plex projects utilize! 

In the case it doesnt work as expected, thankfully it's GPL2.x, so any code they've added I can get back in the case of major issues (Yay open source!) 






> But now, there might actually be a way for you to customize arguments as you see fit, even without changing the existing code.
> Meaning, if you look at the worker.js code, you'll see these two lines near the top:
> 
> ```
> const TRANSCODER_PATH = process.env.TRANSCODER_PATH || '/usr/lib/plexmediaserver/'
> const TRANSCODER_NAME = process.env.TRANSCODER_NAME || 'Plex Transcoder'
> ```
> 
> So technically you can provide your own path and executable name for an ffmpeg transcoder, and the worker will call it, with the original arguments, environment variables, and current working directory value.

Wow! Thanks for that! apreciate the insight! 

I'll probably be going with the first option- because of how "Unique" the jetson deals with ffmpeg, the executable has to be heavily modified to work correctly with different hardware accelerators. 



> Disclaimer: I can't really test these alternatives at the moment, since I run this on a RPI4 cluster within docker swarm without any HW transcoding.

You might be able to do so! I'm just waiting for my boards to arrive in the mail before I can get this up and running; but adding in -hwaccel auto will tell ffmpeg/plex Transcoder to use hardware accelerators if possible! 



> Do you think that might be useful? At least for an initial test?

Very much so! I've build out as much code/infrastructure as I can without having the actual hardware I need to test, so will be doing so when I can!

**Comment by pabloromeo** [comment]
Excellent! Good luck with the Jetson experiment :) Looking forward to seeing how it performed.
